// minecraft-colors/index.js
var colors = {
    0: "\x1b[38;2;0;0;0m",
    1: "\x1b[38;2;0;0;170m",
    2: "\x1b[38;2;0;170;0m",
    3: "\x1b[38;2;0;170;170m",
    4: "\x1b[38;2;170;0;0m",
    5: "\x1b[38;2;170;0;170m",
    6: "\x1b[38;2;255;170;0m",
    7: "\x1b[38;2;170;170;170m",
    8: "\x1b[38;2;85;85;85m",
    9: "\x1b[38;2;85;85;255m",
    a: "\x1b[38;2;85;255;85m",
    b: "\x1b[38;2;85;255;255m",
    c: "\x1b[38;2;255;85;85m",
    d: "\x1b[38;2;255;85;255m",
    e: "\x1b[38;2;255;255;85m",
    f: "\x1b[38;2;255;255;255m",
    k: "\x1b[5m",
    l: "\x1b[1m",
    m: "\x1b[9m",
    n: "\x1b[4m",
    o: "\x1b[3m",
    r: "\x1b[0m"
};
var stripColors = function(str){
    var reg =/\u00A7[0-9|a-z]/gi, res, ind = [], newtext = str.slice(0);
    while((res=reg.exec(str))){ind.push(res.index)};
    for(var i=0; i<ind.length; i++){
        var code = str.slice(ind[i],ind[i]+2);
        newtext=newtext.replace(code, '');
    };
    return newtext;
}
var translateColors = function(str){
    var reg =/\u00A7[0-9|a-z]/gi, res, ind = [], newtext = str.slice(0);
    while((res=reg.exec(str))){ind.push(res.index)};
    for(var i=0; i<ind.length; i++){
        var code = str.slice(ind[i],ind[i]+2);
        newtext=newtext.replace(code, colors[code.replace('\u00A7', '')]);
    };
    return newtext+colors.r;
}

module.exports.colors = colors;
module.exports.stripColors = stripColors;
module.exports.translateColors = translateColors;
